contracts/SponsorFaucet.sol

```diff
@@ -131,95 +97,80 @@ contract SponsorFaucet is Ownable, Pausable, ReentrancyGuard, WhitelistedRole {
-        uint256 gasBound,
-        uint256 collateralBound,
-        uint256 upperBound
+        bounds memory customBounds
     ) public {
         _onlyWhitelistAdmin();
         //rule by internal contract
-        require(upperBound.mul(1000) <= gasBound, "upperBound too high");
-        special_dapps[addr].gas_total_limit = gasTotalLimit;
-        special_dapps[addr].collateral_total_limit = collateralTotalLimit;
-        special_dapps[addr].gas_bound = gasBound;
-        special_dapps[addr].collateral_bound = collateralBound;
-        special_dapps[addr].upper_bound = upperBound;
+        require(customBounds.upper_bound.mul(1000) <= customBounds.gas_bound, "upperBound too high");
+        dapp_bounds[addr] = customBounds;
     }
 
 
     /* ===== Public utility functions ===== */
-    function addGeneralContracts(address[] memory addAddr) public {
+    function addLargeContracts(address[] memory addAddr) public {
         _onlyWhitelistAdmin();
         for(uint256 i = 0; i < addAddr.length; i++) {
-            general_contracts.push(addAddr[i]);
+            large_contracts[addAddr[i]] = true;
+            if(custom_contracts[addAddr[i]]) {
+                delete dapp_bounds[addAddr[i]];
+            }
+            custom_contracts[addAddr[i]] = false;
         }
     }
 
-    function addSpecialContracts(address[] memory addAddr) public {
+    function addCustomContracts(address[] memory addAddr, bounds[] memory addrBounds) public {
         _onlyWhitelistAdmin();
+        require(addAddr.length == addrBounds.length, "length not match");
         for(uint256 i = 0; i < addAddr.length; i++) {
-            special_contracts.push(addAddr[i]);
+            custom_contracts[addAddr[i]] = true;
+            large_contracts[addAddr[i]] = false;
+            dapp_bounds[addAddr[i]] = addrBounds[i];
         }
     }
 
-    function specialContractsCount() public view returns (uint256) {
-        return special_contracts.length;
+    function removeLargeContract(address[] memory removeAddr) public {
+        _onlyWhitelistAdmin();
+        for(uint256 i = 0; i < removeAddr.length; i++) {
+            delete large_contracts[removeAddr[i]];
+            delete dapp_bounds[removeAddr[i]];
+        }
     }
 
-    function generalContractsCount() public view returns (uint256) {
-        return general_contracts.length;
+    function removeCustomContract(address[] memory removeAddr) public {
+        _onlyWhitelistAdmin();
+        for(uint256 i = 0; i < removeAddr.length; i++) {
+            delete custom_contracts[removeAddr[i]];
+            delete dapp_bounds[removeAddr[i]];
+        }
     }
 
-    function isGeneralContract(address dapp) public view returns (bool) {
-        for(uint256 i = 0; i < general_contracts.length; i++) {
-            if(general_contracts[i] == dapp) return true;
-        }
-        return false;
+    function isLargeContract(address dapp) public view returns (bool) {
+        return large_contracts[dapp];
```

Comment:
If getter provided, then suggest to mark the `large_contracts` as private.

Reply:
modified